### PR TITLE
Cleanup PATH variable mangling

### PR DESCRIPTION
**Pull request by jerome-benoit**
### Checklist
- [x] Code compiles correctly
- [x] Relevant tests were added (unit / contract / integration)
- [x] Relevant logs were added
- [x] Formatting and linting run locally successfully
- [x] All tests pass
- [x] UA review
- [x] Design is documented
- [x] Extended the README / documentation, if necessary
- [x] Open source is approved

### CODE REVIEW

File: Dockerfile_mbtci_template [hunk 1]
```diff
@@ -212,6 +212,12 @@ RUN set -ex \
   && python2.7 --version \
   && python3 --version
 
+# Allow global npm packages install without sudo
+RUN set -ex \
+  && mkdir ${USER_HOME_DIR}/.npm-global \
+  && chown ${USER}:${USER} ${USER_HOME_DIR}/.npm-global
```
Comment (jerome-benoit) [inline Dockerfile_mbtci_template:218]:
```suggestion
  && chown -R ${USER}:${USER} ${USER_HOME_DIR}
```

File: Dockerfile_mbtci_template [hunk 2]
```diff
@@ -212,6 +212,12 @@ RUN set -ex \
   && python2.7 --version \
   && python3 --version
 
+# Allow global npm packages install without sudo
+RUN set -ex \
+  && mkdir ${USER_HOME_DIR}/.npm-global \
```
[the remark this answers is not in the document]
Comment (jerome-benoit) [reply Dockerfile_mbtci_template]: Cut&paste  typo ... 

Thanks.

File: Dockerfile_mbtci_template [hunk 3]
```diff
@@ -212,6 +212,12 @@ RUN set -ex \
   && python2.7 --version \
   && python3 --version
 
+# Allow global npm packages install without sudo
+RUN set -ex \
+  && mkdir ${MTA_USER_HOME}/.npm-global \
+  && chown -R ${MTA_USER}:${MTA_USER} ${MTA_USER_HOME}
+ENV NPM_CONFIG_PREFIX ${MTA_USER_HOME}/.npm-global
```
[the remark this answers is not in the document]
Comment (jerome-benoit) [reply Dockerfile_mbtci_template:219]: Both are working. The current Dockerfile use space, so let's keep it consistent with the style already used.

File: Dockerfile_mbtci_template [hunk 4]
```diff
@@ -212,6 +212,12 @@ RUN set -ex \
...
+
 ENV PATH=$PATH:./node_modules/.bin HOME=${MTA_USER_HOME}
```
[the remark this answers is not in the document]
Comment (jerome-benoit) [reply Dockerfile_mbtci_template:221]: It's not intended and a genuine bug, and as well as the HOME variable redefinition. 
The PATH env variable mangling is buggy as hell in that Dockerfile. I plan to fix it but in another PR.

**Comment by jerome-benoit** [comment]
> Does this PR solve #1018 ?

As stated in the PR description, yes, but for currently supported node.js/java versions combination. It does not include node 14 and java 8. 

> How does this PR relate to #1023 ?

That PR is a duplicate of https://github.com/SAP/cloud-mta-build-tool/pull/1013 unrelated to the issue fixed here. 
It's relevant for you as it reintroduces the support for the version combination you are using. 
The original image with the issue have been built with the previous Dockerfile before its rewrite and has serious security issues. In the meantime, you can use the one you've successfully tested to work that introduce the fix in that PR until the java 8 support in re-added and that fix goes in. It does not contain the insecure `chmod -R 777` usage in #1023.

**Comment by rodibrin** [comment]
> As stated in the PR description, yes, but for currently supported node.js/java versions combination. It does not include node 14 and java 8.

Which node/java versions does it cover. we offer the following to the customer:

- devxci/mbtci-java11-node14:
- devxci/mbtci-java11-node16:
- devxci/mbtci-java11-node18:
- devxci/mbtci-java8-node16: 
- devxci/mbtci-java8-node18: 
- devxci/mbtci-java8-node14: 

>  In the meantime, you can use the one you've successfully tested to ...

Unfortunately, the CICD service doesn't allow to easily switch to an arbitrary image. So we have to wait for the release of #1023, right?

Out of curiosity: are there any test processes running I could have a look into?

**Comment by jerome-benoit** [comment]
> 
> devxci/mbtci-java11-node14:
> devxci/mbtci-java11-node16:
> devxci/mbtci-java11-node18:
> —
> 
> 

Only Java 8 based images are not yet shipped as images generated the latest Dockerfile in that repo.

--
Jerome BENOIT - SAP E-Mobility R&D Security and Software Engineer
SAP S/4HANA IC & Q2C ENG Europe III
SAP Labs France SAS, BP1216,805 Av. du Dr. Maurice Donat, 06254 Mougins Cedex, France

***@***.*** - +33 6 35 50 78 40
OpenPGP Key ID : 27B535D3
Key fingerprint : B799 BBF6 8EC8 911B B8D7 CDBC C3B1 92C6 27B5 35D3

Please consider the impact on the environment before printing this email.

**Review by rodibrin — APPROVED** [review]
LGTM

**Comment by jerome-benoit** [comment]
> LGTM

@young-yang03: I think the `HOME` variable override is causing issues at creating the directory structure under `NPM_CONFIG_PREFIX` path. I will remove it in that PR.

Since that PR can't cause any regressions as as long as npm -g is used, it's fine to merge them, ship the images and let @rodibrin confirm it works.

File: Dockerfile_mbtci_template (hunk 4 above)
Comment (jerome-benoit) [inline Dockerfile_mbtci_template:221]:
```suggestion
ENV PATH=$PATH:./node_modules/.bin
```

**Comment by jerome-benoit** [comment]
> LGTM

@rodibrin Have you also tested the node 16 java variant: https://hub.docker.com/repository/docker/fraggle0/mbt-node16-java8-docker/general? 
If not, could you please confirm (or infirm) it works?

**Comment by rodibrin** [comment]
Test of 
`fraggle0/mbt-node16-java8-docker:latest 0f64f79118ce   9 days ago      665MB`
failed:
```
npm ERR! code 254
npm ERR! path /project/node_modules/hana-cli
npm ERR! command failed
npm ERR! command sh -c -- npm link @sap/cds-dk --local
npm ERR! npm ERR! code ENOENT
npm ERR! npm ERR! syscall lstat
npm ERR! npm ERR! path /home/mta/.npm-global/lib
npm ERR! npm ERR! errno -2
npm ERR! npm ERR! enoent ENOENT: no such file or directory, lstat '/home/mta/.npm-global/lib'
npm ERR! npm ERR! enoent This is related to npm not being able to find a file.
npm ERR! npm ERR! enoent
npm ERR!
npm ERR! npm ERR! A complete log of this run can be found in:
npm ERR! npm ERR!     /home/mta/.npm/_logs/2023-02-02T09_12_44_959Z-debug-0.log

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/mta/.npm/_logs/2023-02-02T08_55_30_802Z-debug-0.log
```

```
mta@ba01bd395d75:/project$ ls -ali /home/mta/
total 48
1926508 drwxrwxrwx 1 mta  mta  4096 Feb  2 09:12 .
1926507 drwxr-xr-x 1 root root 4096 Jan 16 22:47 ..
1926509 -rw-r--r-- 1 mta  mta   220 Mar 27  2022 .bash_logout
1926510 -rw-r--r-- 1 mta  mta  3526 Mar 27  2022 .bashrc
1934075 drwxr-xr-x 3 mta  mta  4096 Feb  2 09:12 .cache
1934070 drwx------ 3 mta  mta  4096 Feb  2 09:12 .hdb
1926511 drwxr-xr-x 1 mta  mta  4096 Feb  2 08:55 .npm
1926517 drwxr-xr-x 2 mta  mta  4096 Jan 23 10:52 .npm-global
1926518 -rw-r--r-- 1 mta  mta   807 Mar 27  2022 .profile
mta@ba01bd395d75:/project$ ls -ali /home/mta/.npm-global/
total 12
1926517 drwxr-xr-x 2 mta mta 4096 Jan 23 10:52 .
1926508 drwxrwxrwx 1 mta mta 4096 Feb  2 09:12 ..

```

**Comment by jerome-benoit** [comment]
> Test of `fraggle0/mbt-node16-java8-docker:latest 0f64f79118ce 9 days ago 665MB` failed:

Ok, thanks. I'm pushing the test images with the directory pre-creation. Dunno why npm does not properly populate it with later version.

Could you please re-test?

File: Dockerfile_mbtci_template [hunk 5]
```diff
@@ -212,6 +212,12 @@ RUN set -ex \
   && python2.7 --version \
   && python3 --version
 
-ENV PATH=$PATH:./node_modules/.bin HOME=${MTA_USER_HOME}
+# Allow global npm packages install without sudo
+RUN set -ex \
+  && mkdir ${MTA_USER_HOME}/.npm-global \
```
Comment (jerome-benoit) [inline Dockerfile_mbtci_template:217]:
```suggestion
  && mkdir ${MTA_USER_HOME}/.npm-global \
  && mkdir ${MTA_USER_HOME}/.npm-global/lib \
```

**Comment by rodibrin** [comment]
test succeeded:

`fraggle0/mbt-node16-java8-docker   latest    bcc6a780a618   3 hours ago    665MB`

**Review by young-yang03 — APPROVED** [review]

**Pull request closed without merging**